kimchi/src/circuits/polynomials/generic.rs

```diff
@@ -1,73 +1,284 @@
+    pub fn create_generic_easy(
+        wires: GateWires,
+        gate1: GenericGate<F>,
+        gate2: Option<GenericGate<F>>,
+    ) -> Self {
+        let mut coeffs = [F::zero(); GENERICS * 2 + 2 + 2];
+        match gate1 {
+            GenericGate::Add {
+                left_coeff,
+                right_coeff,
+                output_coeff,
+            } => {
+                coeffs[0] = left_coeff.unwrap_or_else(F::one);
+                coeffs[1] = right_coeff.unwrap_or_else(F::one);
+                coeffs[2] = output_coeff.unwrap_or_else(|| -F::one());
+            }
+            GenericGate::Mul {
+                output_coeff,
+                mul_coeff,
+            } => {
+                coeffs[2] = output_coeff.unwrap_or_else(|| -F::one());
+                coeffs[3] = mul_coeff.unwrap_or_else(F::one);
+            }
+            GenericGate::Const(cst) => {
+                coeffs[0] = F::one();
+                coeffs[4] = -cst;
+            }
+            GenericGate::Pub => {
+                coeffs[0] = F::one();
+            }
+        };
+        match gate2 {
+            Some(GenericGate::Add {
+                left_coeff,
+                right_coeff,
+                output_coeff,
+            }) => {
+                coeffs[5] = left_coeff.unwrap_or_else(F::one);
+                coeffs[6] = right_coeff.unwrap_or_else(F::one);
+                coeffs[7] = output_coeff.unwrap_or_else(|| -F::one());
+            }
+            Some(GenericGate::Mul {
+                output_coeff,
+                mul_coeff,
+            }) => {
+                coeffs[7] = output_coeff.unwrap_or_else(|| -F::one());
+                coeffs[8] = mul_coeff.unwrap_or_else(F::one);
+            }
+            Some(GenericGate::Const(cst)) => {
+                coeffs[5] = F::one();
+                coeffs[9] = -cst;
+            }
+            Some(GenericGate::Pub) => {
+                coeffs[5] = F::one();
+                unimplemented!();
+            }
+            None => (),
+        };
+        Self::create_generic(wires, coeffs)
+    }
+
+    /// verifies that the generic gate constraints are solved by the witness
+    pub fn verify_generic(&self, row: usize, witness: &[Vec<F>; COLUMNS]) -> Result<(), String> {
+        // assignments
+        let this: [F; COLUMNS] = array_init(|i| witness[i][row]);
+
+        // constants
+        let zero = F::zero();
+
+        // check if it's the correct gate
+        ensure_eq!(self.typ, GateType::Generic, "generic: incorrect gate");
+
+        // toggling each column x[i] depending on the selectors c[i]
+        let sum1 = (0..GENERICS)
+            .map(|i| self.coeffs[i] * this[i])
+            .fold(zero, |x, y| x + y);
+
+        let sum2 = (0..GENERICS)
+            .map(|i| self.coeffs[i + GENERICS_COEFFS] * this[i + GENERICS])
+            .fold(zero, |x, y| x + y);
+
+        // multiplication
+        let mul1 = self.coeffs[3] * this[0] * this[1];
```

Comment:
Similar to what Matthew said before, I think it is less error prone if we assign constants to these indices